javascript/ql/lib/semmle/javascript/frameworks/data/internal/Shared.qll

```diff
@@ -0,0 +1,613 @@
+    row.splitAt(";", 1) = type and
+    normalizePath(row.splitAt(";", 2)) = path and
+    row.splitAt(";", 3) = kind
+  )
+}
+
+/** Holds if a sink model exists for the given parameters. */
+private predicate sinkModel(string package, string type, string path, string kind) {
+  exists(string row |
+    sinkModel(row) and
+    row.splitAt(";", 0) = package and
+    row.splitAt(";", 1) = type and
+    normalizePath(row.splitAt(";", 2)) = path and
+    row.splitAt(";", 3) = kind
+  )
+}
+
+/** Holds if a summary model `row` exists for the given parameters. */
+private predicate summaryModel(
+  string package, string type, string path, string input, string output, string kind
+) {
+  exists(string row |
+    summaryModel(row) and
+    row.splitAt(";", 0) = package and
+    row.splitAt(";", 1) = type and
+    normalizePath(row.splitAt(";", 2)) = path and
+    normalizePath(row.splitAt(";", 3)) = input and
+    normalizePath(row.splitAt(";", 4)) = output and
+    row.splitAt(";", 5) = kind
+  )
+}
+
+/** Holds if an type model exists for the given parameters. */
+private predicate typeModel(
+  string package1, string type1, string package2, string type2, string path
+) {
+  exists(string row |
+    typeModel(row) and
+    row.splitAt(";", 0) = package1 and
+    row.splitAt(";", 1) = type1 and
+    row.splitAt(";", 2) = package2 and
+    row.splitAt(";", 3) = type2 and
+    normalizePath(row.splitAt(";", 4)) = path
+  )
+}
+
+/**
+ * Gets a package that should be seen as an alias for the given other `package`,
+ * or the `package` itself.
+ */
+bindingset[package]
+bindingset[result]
+string getAPackageAlias(string package) {
+  typeModel(package, "", result, "", "")
+  or
+  result = package
+}
+
+/**
+ * Holds if CSV rows involving `package` might be relevant for the analysis of this database.
+ */
+private predicate isRelevantPackage(string package) {
+  Impl::isPackageUsed(package)
+  or
+  exists(string other |
+    isRelevantPackage(other) and
+    typeModel(package, _, other, _, _)
+  )
+}
+
+/**
+ * Holds if `package,type,path` is used in some CSV row.
+ */
+pragma[nomagic]
+predicate isRelevantFullPath(string package, string type, string path) {
+  isRelevantPackage(package) and
+  (
+    sourceModel(package, type, path, _) or
+    sinkModel(package, type, path, _) or
+    summaryModel(package, type, path, _, _, _) or
+    typeModel(_, _, package, type, path)
+  )
+}
+
+/**
+ * Holds if `path` or some suffix thereof is used with the `package,type` combination in some CSV row.
+ */
+pragma[nomagic]
+predicate isRelevantPath(string package, string type, string path) {
+  exists(string fullPath |
+    isRelevantFullPath(package, type, fullPath) and
+    path = fullPath.prefix([0, fullPath.indexOf("."), fullPath.length()])
+  )
+}
+
+/** Holds if `path` has the form `basePath.token` where `token` is a single token. */
+bindingset[path]
+private predicate decomposePath(string path, string basePath, string token) {
```

Comment:
This way of decomposing is quadratic in the number of tokens in `path`. Might as well use a linear decomposition. I.e. for `"a.b.c.d"` just construct each of `"a"`, `"b"`, `"c"`, and `"d"`, instead of also constructing `"a.b.c"` and `"a.b"`.

Reply:
Probably doesn't matter because of the package-level pruning, but yeah we might as well clean it up.

We previously needed a value to represent each step of the access path, which we don't get by simply splitting into individual tokens. The resolution predicates now take an `int n` parameter and resolve the first `n` tokens of the given full access path, with no sharing of common prefixes.

To simplify accessing the tokens of an access path I added the `AccessPath` class as well.